src/editor/mod.rs

```diff
@@ -204,48 +204,35 @@ impl Editor {
     }
 
     fn scroll_region(&mut self, top: u64, bot: u64, left: u64, right: u64, rows: i64, cols: i64) {
-        let (top, bot) =  if rows > 0 {
-            (top as i64 + rows, bot as i64)
-        } else if rows < 0 {
-            (top as i64, bot as i64 + rows)
+        let (mut y, end_y, step_y) = if rows > 0 {
+            (top as i64 + rows, bot as i64, 1)
         } else {
-            (top as i64, bot as i64)
+            (bot as i64 + rows - 1, top as i64 - 1, -1)
         };
 
-        let (left, right) = if cols > 0 {
-            (left as i64 + cols, right as i64)
-        } else if rows < 0 {
-            (left as i64, right as i64 + cols)
+        let (start_x, end_x, step_x) = if cols > 0 {
+            (left as i64 + cols, right as i64, 1)
         } else {
-            (left as i64, right as i64)
+            (right as i64 + cols - 1, left as i64 - 1, -1)
         };
 
-        let mut region = Vec::new();
-        for y in top..bot {
-            let row = &self.grid[y as usize];
-            let mut copied_section = Vec::new();
-            for x in left..right {
-                copied_section.push(row[x as usize].clone());
-            }
-            region.push(copied_section);
-        }
-
-        let new_top = top as i64 - rows;
-        let new_left = left as i64 - cols;
-
-        for (y, row_section) in region.into_iter().enumerate() {
-            for (x, cell) in row_section.into_iter().enumerate() {
-                let y = new_top + y as i64;
-                if y >= 0 && y < self.grid.len() as i64 {
-                    let row = &mut self.grid[y as usize];
-                    let dirty_row = &mut self.dirty[y as usize];
-                    let x = new_left + x as i64;
-                    if x >= 0 && x < row.len() as i64 {
-                        row[x as usize] = cell;
-                        dirty_row[x as usize] = true;
+        while y != end_y {
```

Comment:
My only real hesitation here is that these while loops feel like they are itching to be for loops. Other than that this looks great.

Reply:
Reimplemented as a ranged for loop. Not terribly happy about the allocations, but `rev()` returning an entirely different type forced me to use a `Box`, and I haven't been able to find a cleaner way

Reply:
I'll spend some time thinking if there is something cleaner. Merged for noq